### PR TITLE
Mithril relay broadcasts signer registrations in P2P

### LINKED ISSUE #1587: Mithril relay broadcasts signer registrations with P2P PubSub
## Why
We want to keep experimenting with the P2P relay that we have prototyped to broadcast signer registrations in a similar way that we already broadcast signature registrations. 

> [!NOTE]
> This is only an experimentation on the Mithril relay which is not production ready yet in order to decentralize further Mithril networks. We already know that there is no consensus on the P2P network so we expect this experiment to work in an ideal adversary free environment.

## What
The Mithril relay should handle the following behavior: 
- The relay on the signer side should broadcast messages on a P2P network 
- The relay on the aggregator side should catch these messages and forward them to the aggregator with a HTTP call

## How

- [x] Mithril relay broadcasts signer registrations in an ideal adversary free environment
- [x] Mithril relay repeats signer registrations in order to minimize the expected on-boarding delay of new/recently offline signer/aggregator
- [x] Enhancements of the relay
  - [x]  Fix CI flakiness due to P2P latency vs end to end test parameters
  - [x]  Fix slow test `should_receive_signatures_from_signers_when_subscribed_to_pubsub`
  - [x]  Implement peer signature sent to aggregator retry

## Parent issue
#1456

### PR DESCRIPTION
**Pull request by jpraynaud**
## Content
This PR includes the implementation of a prototypical decentralized signer registration with the P2P network of the Mithril relay:
- signer registration messages are broadcast to the P2P network
- signer registration messages are repeated at regular interval in order to support: 
  - lost messages on the P2P network (e.g. due to network connectivity between peers)
  - newly signer/aggregator joining the network (to avoid `1` more epoch delay before being able to contribute)
  - recently offline signer/aggregator (to avoid `1` more epoch delay before being able to contribute)

## Pre-submit checklist

- Branch
  - [x] Tests are provided (if possible)
  - [x] Crates versions are updated (if relevant)
  - [x] Commit sequence broadly makes sense
  - [x] Key commits have useful messages
- PR
  - [x] No clippy warnings in the CI
  - [x] Self-reviewed the diff
  - [x] Useful pull request description
  - [x] Reviewer requested

## Issue(s)
Closes #1587

### CODE REVIEW

File: mithril-relay/Cargo.toml [hunk 1]
```diff
@@ -46,5 +46,5 @@ slog-bunyan = "2.5.0"
 slog-scope = "4.4.0"
 slog-term = "2.9.0"
 thiserror = "1.0.56"
-tokio = { version = "1.35.1", features = ["full"] }
+tokio = { version = "1.37.0", features = ["full"] }
```
Comment (dlachaume) [inline mithril-relay/Cargo.toml:49]: Should it be updated in the other crates as well? Or would it make more sense to do it in a separate PR?

File: mithril-relay/src/repeater.rs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+use mithril_common::StdResult;
+use slog_scope::debug;
+use std::{fmt::Debug, sync::Arc, time::Duration};
+use tokio::{
+    sync::{mpsc::UnboundedSender, Mutex},
+    time::Instant,
+};
+
+/// A message repeater will send a copy of the message to a channel at a given frequency
+pub struct MessageRepeater<M: Clone + Debug + Sync + Send + 'static> {
```
Comment (dlachaume) [inline mithril-relay/src/repeater.rs:11]: Given the size of `RegisterSignerMessage`, could it be a problem to clone the message every time?

File: mithril-relay/src/commands/signer.rs [hunk 3]
```diff
@@ -32,8 +38,15 @@ impl SignerCommand {
         let dial_to = self.dial_to.to_owned();
         let addr: Multiaddr = format!("/ip4/0.0.0.0/tcp/{}", self.listen_port).parse()?;
         let aggregator_endpoint = self.aggregator_endpoint.to_owned();
-
-        let mut relay = SignerRelay::start(&addr, &server_port, &aggregator_endpoint).await?;
+        let signer_repeater_delay = Duration::from_millis(self.signer_repeater_delay.to_owned());
```
Comment (dlachaume) [inline mithril-relay/src/commands/signer.rs:41]:
```suggestion
        let signer_repeater_delay = Duration::from_millis(self.signer_repeater_delay);
```

**Review by dlachaume — APPROVED** [review]
LGTM 👍. Just some suggestions/questions and typo to fix.

File: mithril-relay/Cargo.toml (hunk 1 above)
Comment (jpraynaud) [reply mithril-relay/Cargo.toml:49]: I updated for the relay as it was needed for the latest version of `libp2p`.
We should definitely update the other crates it in a separate PR (I added a task in #1614 for that).

File: mithril-relay/src/p2p/peer.rs [hunk 4]
```diff
@@ -189,31 +211,60 @@ impl Peer {
...
+                format!(
+                    "Can not publish broadcast message on {topic_name} P2P pubsub without swarm"
+                )
+            })?;
+
         Ok(message_id.to_owned())
     }
 
+    /// Publish a signer registration on the P2P pubsub
+    pub fn publish_signer(
```
[the remark this answers is not in the document]
Comment (jpraynaud) [reply mithril-relay/src/p2p/peer.rs:258]: Done

File: mithril-relay/src/repeater.rs (hunk 2 above)
Comment (jpraynaud) [reply mithril-relay/src/repeater.rs:11]: If you mean cloning the message sent to the channel, I guess that is the expected way of passing messages to a channel.
Also, I don't really foresee any problem with the size as it's only few KB.

Can you further elaborate your concerns?

File: mithril-relay/tests/register_signer_signature.rs [hunk 5]
```diff
@@ -1,15 +1,20 @@
...
+    PassiveRelay, SignerRelay,
+};
 use reqwest::StatusCode;
 use slog::{Drain, Level, Logger};
-use slog_scope::info;
+use slog_scope::{error, info};
 
 // Launch a relay that connects to P2P network. The relay is a peer in the P2P
-// network. The relay sends some signatures that must be received by other
+// network. The relay sends some signer regsitrations that must be received by other
```
[the remark this answers is not in the document]
Comment (jpraynaud) [reply mithril-relay/tests/register_signer_signature.rs:14]: Done

File: mithril-relay/src/relay/aggregator.rs [hunk 6]
```diff
@@ -48,25 +51,68 @@ impl AggregatorRelay {
...
+            },
+            Err(err) => {
+                error!("Relay aggregator: Post `/register-signer` failed: {err:?}");
+                Err(anyhow!("Post `/register-signer` failed: {err:?}"))
+            }
+        }
+    }
+
     /// Tick the aggregator relay
     pub async fn tick(&mut self) -> StdResult<()> {
```
[the remark this answers is not in the document]
Comment (jpraynaud) [reply mithril-relay/src/relay/aggregator.rs:83]: I guess that it's going to be a nightmare to implement and produce code which is not very readbale.
We should keep it like this IMO.

**Review by sfauvel — APPROVED** [review]
Looks good

File: mithril-relay/src/relay/signer.rs [hunk 7]
```diff
@@ -23,24 +28,41 @@ impl SignerRelay {
...
         server_port: &u16,
         aggregator_endpoint: &str,
+        signer_repeater_delay: &Duration,
     ) -> StdResult<Self> {
         debug!("SignerRelay: starting...");
         let (signature_tx, signature_rx) = unbounded_channel::<RegisterSignatureMessage>();
+        let (signer_tx, signer_rx) = unbounded_channel::<RegisterSignerMessage>();
+        let signer_repeater = Arc::new(MessageRepeater::new(
+            signer_tx.clone(),
+            signer_repeater_delay.to_owned(),
```
[the remark this answers is not in the document]
Comment (dlachaume) [reply mithril-relay/src/relay/signer.rs:38]: If you only write `signer_repeater_delay`, you will have this suggestion.

Screenshot 2024-04-17 at 18 07 28

File: mithril-relay/src/commands/signer.rs [hunk 8]
```diff
@@ -23,6 +25,10 @@ pub struct SignerCommand {
     /// Aggregator endpoint URL.
     #[clap(long, env = "AGGREGATOR_ENDPOINT")]
     aggregator_endpoint: String,
+
+    /// Signer registration repeater delay in milliseconds (defaults to 1 hour)
```
Comment (Alenar) [inline mithril-relay/src/commands/signer.rs:29]:
```suggestion
    /// Interval at which a Signer registration should be repeated in milliseconds (defaults to 1 hour)
```

File: mithril-relay/src/p2p/peer.rs [hunk 9]
```diff
@@ -75,28 +89,36 @@ impl Peer {
     }
 
     fn build_topics() -> HashMap<TopicName, gossipsub::IdentTopic> {
-        HashMap::from([(
-            MITHRIL_SIGNATURES_TOPIC_NAME.into(),
-            gossipsub::IdentTopic::new(MITHRIL_SIGNATURES_TOPIC_NAME),
-        )])
+        HashMap::from([
+            (
+                MITHRIL_SIGNATURES_TOPIC_NAME.into(),
```
Comment (Alenar) [inline mithril-relay/src/p2p/peer.rs:94]: Maybe the topic could be gathered in a mod (something like that):
```rust
pub mod mithril_p2p_topic {
  pub const SIGNATURES: &str = "mithril/signatures";
  pub const SIGNERS: &str = "mithril/signers";
}
```

**Review by Alenar — APPROVED** [review]

File: mithril-relay/src/commands/signer.rs (hunk 8 above)
Comment (jpraynaud) [reply mithril-relay/src/commands/signer.rs:29]: Done

File: mithril-relay/src/p2p/peer.rs (hunk 9 above)
Comment (jpraynaud) [reply mithril-relay/src/p2p/peer.rs:94]: Done

File: mithril-relay/src/relay/aggregator.rs [hunk 10]
```diff
@@ -48,25 +51,68 @@ impl AggregatorRelay {
...
     }
 
+    async fn notify_signer_to_aggregator(
+        &self,
+        signer_message: &RegisterSignerMessage,
+    ) -> StdResult<()> {
+        let response = reqwest::Client::new()
+            .post(format!("{}/register-signer", self.aggregator_endpoint))
+            .json(signer_message)
+            //.header(MITHRIL_API_VERSION_HEADER, "0.1.13") // TODO: retrieve current version
```
[the remark this answers is not in the document]
Comment (jpraynaud) [reply mithril-relay/src/relay/aggregator.rs:61]: The handling of the API version is an open question as it get tricky when handling Mithril era transitions. Let's keep the comment here until we have a better idea of what to do

File: mithril-relay/src/relay/aggregator.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jpraynaud) [reply mithril-relay/src/relay/aggregator.rs:83]: I made an attempt here: https://github.com/input-output-hk/mithril/commit/9d2113d7bdb5b6cce2f7924c0dacd9fc27830f46

But it's not successful, because of async code.

If you want to give it a try, feel free to update the branch :slightly_smiling_face:
